core/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourceSubPhase.java

```diff
@@ -35,29 +42,50 @@ public void hitExecute(SearchContext context, HitContext hitContext) {
         if (context.sourceRequested() == false) {
             return;
         }
+        boolean rootHit = hitContext.hit().getNestedIdentity() == null;
         SourceLookup source = context.lookup().source();
         FetchSourceContext fetchSourceContext = context.fetchSourceContext();
         assert fetchSourceContext.fetchSource();
         if (fetchSourceContext.includes().length == 0 && fetchSourceContext.excludes().length == 0) {
-            hitContext.hit().sourceRef(source.internalSourceRef());
+            if (rootHit ) {
+                hitContext.hit().sourceRef(source.internalSourceRef());
+            } else {
+                try {
+                    applyNestedSource(source.source(), source.sourceContentType(), hitContext);
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
             return;
         }
 
-        if (source.internalSourceRef() == null) {
+        if (rootHit && source.internalSourceRef() == null) {
             throw new IllegalArgumentException("unable to fetch fields from _source field: _source is disabled in the mappings " +
                     "for index [" + context.indexShard().shardId().getIndexName() + "]");
         }
 
         final Object value = source.filter(fetchSourceContext);
         try {
-            final int initialCapacity = Math.min(1024, source.internalSourceRef().length());
-            BytesStreamOutput streamOutput = new BytesStreamOutput(initialCapacity);
-            XContentBuilder builder = new XContentBuilder(source.sourceContentType().xContent(), streamOutput);
-            builder.value(value);
-            hitContext.hit().sourceRef(builder.bytes());
+            if (rootHit) {
+                final int initialCapacity = Math.min(1024, source.internalSourceRef().length());
+                BytesStreamOutput streamOutput = new BytesStreamOutput(initialCapacity);
```

Comment:
It looks to me like it duplicates the logic of creating a XContentBuilder in a given type and then write the filtered source as map. Could it be something like this?

``` 
...
Object value = source.filter(fetchSourceContext);
try {
    if (nestedHit) {
        value = getNestedSource((Map<String, Object>) value, hitContext);
    }
    final int initialCapacity = Math.min(1024, source.internalSourceRef().length()); // deal with null here
    try (BytesStreamOutput streamOutput = new BytesStreamOutput(initialCapacity)) {
        XContentBuilder builder = new XContentBuilder(source.sourceContentType().xContent(), streamOutput);
        builder.value(value);
        hitContext.hit().sourceRef(builder.bytes());
    }
...
```